src/api/oauth_routes.ts

```diff
@@ -0,0 +1,474 @@
+        // OAuth application management
+        app.get("/oauth/applications/create", csurf(), LocalAuthMiddleware, this.createOAuthApplicationPage)
+        app.post("/oauth/applications", csurf(), LocalAuthMiddleware, this.createOAuthApplication)
+        app.get("/oauth/applications", csurf(), LocalAuthMiddleware, this.listOAuthApplications)
+        app.get("/oauth/applications/:applicationID", csurf(), LocalAuthMiddleware, this.oauthApplicationPage)
+        app.put("/oauth/applications/:applicationID", csurf(), LocalAuthMiddleware, this.updateOAuthApplication )
+        app.delete("/oauth/applications/:applicationID", csurf(), LocalAuthMiddleware, this.deleteOAuthApplication)
+
+        // login, register and authorize
+        app.get("/", csurf(), this.loginPage)
+        app.get("/logout", this.logout)
+        app.post("/", csurf(),  LocalAuthMiddleware, this.login)
+
+        app.get("/oauth/register", csurf(), this.registerPage)
+        app.post("/oauth/register", csurf(), this.createNewUser)
+
+        app.get("/oauth/authorize", csurf(), LocalAuthMiddleware, this.authorizePage)
+        app.post("/oauth/authorize", csurf(), LocalAuthMiddleware, this.authorize)
+        app.post("/oauth/exchange", this.tokenExchange)
+
+        // profile management and email validation/reset password
+        app.get("/oauth/profile",csurf(), LocalAuthMiddleware, this.profile)
+        app.post("/oauth/profile/keys", csurf(), LocalAuthMiddleware, this.generateKeyPair)
+        app.delete("/oauth/profile/keys/:keyID", csurf(), LocalAuthMiddleware, this.deleteKeyPair)
+
+        app.get("/validate-email", this.validateEmail)
+
+        app.get("/reset-password",csurf(), this.resetPasswordPage)
+        app.post("/reset-password", csurf(), this.initiatePasswordReset)
+        app.post("/reset-password/reset",csurf(), this.resetPassword)
+    }
+
+    private static profile(req: Request, res: Response, next: NextFunction) {
+        const user = req.user as UserT
+        // profile must include a user's keys
+        KeyPairStorage.Instance.KeysForUser(user.id!)
+            .then((result) => {
+                if (result.isError && result.error) {
+                    res.render('profile', {_error: "Unable to fetch user's API keys"})
+                    return
+                }
+
+                // @ts-ignore
+                res.render('profile', {
+                    // @ts-ignore
+                    _csrfToken: req.csrfToken(),
+                    user: req.user,
+                    apiKeys: result.value,
+                    _success: req.query.success,
+                    _error: req.query.error,
+                    api_key: req.query.newKey,
+                    api_secret: req.query.newSecret
+                })
+                return
+            })
+            .catch((err) => res.render('profile', {_error: err}))
+    }
+
+    private static createOAuthApplicationPage(req: Request, res: Response) {
```

Comment:
Maybe consider renaming this method so that it's clear nothing is being created?

Reply:
text_ops column is missing :)

Reply:
Good to Merge once @jerenmb is addressed and text_ops is fixed

Reply:
Both addressed